src/solvers/file_solution_history.C

```diff
@@ -33,81 +33,73 @@ FileSolutionHistory::~FileSolutionHistory ()
 
 // This function finds, if it can, the entry where we're supposed to
 // be storing data
-void FileSolutionHistory::find_stored_entry()
+void FileSolutionHistory::find_stored_entry(Real time)
 {
   if (stored_solutions.begin() == stored_solutions.end())
     return;
 
   libmesh_assert (stored_sols != stored_solutions.end());
 
-  if (std::abs(stored_sols->first - _system.time) < TOLERANCE)
-    return;
+  // We will use the map::lower_bound operation to find the least key, which
+  // is the least upper bound among all the keys for the time of our interest.
+  // (key before map::lower_bound) < time < map::lower_bound, one of these
+  // will be within TOLERANCE of time (unless we are creating a new map entry)
+  stored_solutions_iterator lower_bound_it = stored_solutions.lower_bound(time);
 
-  // If we're not at the front, check the previous entry
-  if (stored_sols != stored_solutions.begin())
-    {
-      stored_solutions_iterator test_it = stored_sols;
-      if (std::abs((--test_it)->first - _system.time) < TOLERANCE)
-        {
-          --stored_sols;
-          return;
-        }
-    }
+  // Set the stored sols iterator to whichever key is within TOLERANCE of time
+  if(std::abs(lower_bound_it->first - time) < TOLERANCE)
+  {
+    stored_sols = stored_solutions.find(time);
+  }
+  else if(std::abs((--lower_bound_it)->first - time) < TOLERANCE)
+  {
+    stored_sols = stored_solutions.find(time);
+  }
```

Comment:
Should we scream and die here if we don't find anything?  At least a libmesh_assert()?

Reply:
Is it not possible for `--lower_bound_it` to step off the front of the array?

Reply:
Oh, wait a minute, this is just wrong.  If find() was going to work, we could have just used it in the first place!  The point of my lower_bound suggestion was that you can then be sure of finding the nearest two entries, and that you can then *use* one of those entries.  If you don't actually use the entry then if we have even a hint of floating point error stored_sols will get set to stored_solutions.end() and we'll either get an assertion failure at a completely different part of the code (e.g. from the assertion above here at the *next* step) or we'll get a segfault or a garbage value when end() gets dereferenced.

We don't need to revert this PR since the bug didn't happen to trigger anything in CI, but we should get this fixed ASAP.